### ISSUE TITLE
Question: What reform uses for UUID support

### ISSUE THREAD

**Issue by rumyantseva** [go-reform/reform#202]
When my model has a UUID-based field and when I do `reform models` I have the following situation:

1. Reform generates the code but doesn't add anything to imports. Personally, I use `github.com/satori/go.uuid` for uuid. I wonder, if I can add something to reform to be able to support this import automatically. 

2. Reform generates the following method:
```
// SetPK sets record primary key.
func (s *Profile) SetPK(pk interface{}) {
	if i64, ok := pk.(int64); ok {
		s.ID = uuid.UUID(i64)
	} else {
		s.ID = pk.(uuid.UUID)
	}
}
```
As nothing has been added to imports, I wonder which library supports `uuid.UUID(i64)` (as this is not what `satori/go.uuid` can offer, it gives me `cannot convert i64 (type int64) to type uuid.UUID`). I found a way to make my code working:

```
func (s *Profile) SetPK(pk interface{}) {
		s.ID = pk.(uuid.UUID)
}
```

But I wonder if this is a suitable approach and how can I improve the whole story.

**Comment by AlekSi** [comment]
See https://github.com/go-reform/reform/issues/43 and https://github.com/go-reform/reform/issues/56.

**Comment by AlekSi** [comment]
We need more information about PK type.

`HasPK` is currently generated with this template:
```
// HasPK returns true if record has non-zero primary key set, false otherwise.
func (s *{{ .Type }}) HasPK() bool {
    return s.{{ .PKField.Name }} != {{ .TableVar }}.z[{{ .TableVar }}.s.PKFieldIndex]
}
```
It compiles only for types that has `!=` operators defined, and slices do not define it. We could use `!reflect.DeepEqual(s.{{ .PKField.Name }}, {{ .TableVar }}.z[{{ .TableVar }}.s.PKFieldIndex])`, but that might be not so good. With knowledge about PK type, we could use code with `!=` for simple types, `s.{{ .PKField.Name }}.Equal({{ .TableVar }}.z[{{ .TableVar }}.s.PKFieldIndex])` for types defining `Equal(x X) bool`, and `reflect.DeepEqual` as fallback.

`SetPK` is currently generated with this template:
```
// SetPK sets record primary key.
func (s *{{ .Type }}) SetPK(pk interface{}) {
    if i64, ok := pk.(int64); ok {
        s.{{ .PKField.Name }} = {{ .PKField.Type }}(i64)
    } else {
        s.{{ .PKField.Name }} = pk.({{ .PKField.Type }})
    }
}
```

It is only used by when `Record` with zero PK is inserted, PK is autogenerated, and RDBMS support `LastInsertId`: https://github.com/go-reform/reform/blob/c8eea006559b6f50a00404d9f81b191da8fd02cc/querier_commands.go#L97. In that case, `SetPK` argument is always `int64`, but PK field can have another type, for example, `int32` or even `type MyPKType int32`. That's why we handle `int64` differently and try a conversion. With knowledge about the PK type, we can use the type-specific code.

There are two ways to get more information about PK type. The first one is to parse files with `go/types`: https://github.com/go-reform/reform/issues/57 The second is to use some reflection in `init()` function to switch implementations. You are welcome to work on any of them.

**Comment by AlekSi** [comment]
I responded in the other issue, but want to highlight one thing there:
```go
func (s *Profile) SetPK(pk interface{}) {
	s.ID = pk.(uuid.UUID)
}
```
You could replace that code with:
```go
func (s *Profile) SetPK(pk interface{}) {
	panic("should not be used")
}
```

`SetPK` is used only with [`sql.Result.LastInsertId`](https://golang.org/pkg/database/sql/#Result) that only works for integer autogenerated keys.

**Comment by rumyantseva** [comment]
@AlekSi thanks! Actually, I was pretty sure that `pq` doesn't support `LastInsertId` at all:

> pq does not support the LastInsertId() method of the Result type in database/sql. To return the identifier of an INSERT (or UPDATE or DELETE), use the Postgres RETURNING clause with a standard Query or QueryRow call

(from https://godoc.org/github.com/lib/pq, also: https://github.com/lib/pq/blob/51e2106eed1cea199c802d2a49e91e2491b02056/conn.go#L694)

So, do you mean that for the `serial` type `LastInsertId` should work even with `lib/pq`?

**Comment by AlekSi** [comment]
Yes, PostgreSQL uses `INSERT … RETURNING id` syntax. `SetPK` method is never used by reform.

**Comment by rumyantseva** [comment]
@AlekSi what do you mean "never used by reform"? (: https://github.com/go-reform/reform/blob/c8eea006559b6f50a00404d9f81b191da8fd02cc/querier_commands.go#L97

**Comment by AlekSi** [comment]
… for PostgreSQL. Reform always uses the code above that line for PostgreSQL.

**Comment by rumyantseva** [comment]
@AlekSi ok, then I agree :) (but, for example, it's not possible to simply remove `SetPK` method from the code generated for the model)

**Comment by AlekSi** [comment]
That's why this and other tickets are still open.
